advisories/a64325.md

```diff
@@ -0,0 +1,41 @@
+---
+title: Technical Advisory 64325
+summary: Race condition between long running reads and replica removal
+toc: true
+---
+
+Publication date: May 3, 2021
+
+## Description
+
+Cockroach Labs has discovered a race condition, where a long running read request submitted during replica removal may be evaluated on the removed replica, returning an empty result. Common causes of replica removal are range merges and replica rebalancing in the cluster. When a replica is removed there is a small window, where the data is already deleted but a read request will still see it as a valid replica. In this case a read request will always return an empty result for the part of the query that relied on this range.
+
+This is a very rare scenario that can only be encountered, if a number of very unlikely conditions are met:
+
+ - A read request checks that the node is holding the lease and begins processing
+ - The node loses the lease before the read request is able to lock the replica for reading
+ - The read request checks the state of the replica and makes sure it's a alive
+ - The new leaseholder immediately removes the old replica from the old leaseholder, due to a rebalance or range merge.
+ - The removal is processed before the read requests starts reading the data from the replica
+ - The read request observes the replica in a post-deletion state returning an empty result set.
+
+ Due to the nature of this race condition, we believe that it is only likely for this to happen on an extremely overloaded node that is struggling to keep up with processing requests and experiences natural delays between each step in the processing of a read request.
+
+This issue affects CockroachDB [versions](/docs/releases/) 19.2 and later.
+
+## Statement
+This is resolved in CockroachDB by PR [#64324] and [#64471], which fix the race condition between read requests and replica removal.
+
+The fix has been applied to maintenance releases of CockroachDB v20.1, v20.2, and v21.1.
+
+This public issue is tracked by [#64325] and [#46329]
```

Comment:
I wouldn't link to #46329. While it's related, it's not the same issue. In some ways, it's worse, because we weren't even trying to get the synchronization right. But it's also much less severe in its potential outcomes and we have no evidence that anyone has actually run into it.